query-algebrizer/src/clauses/or.rs

```diff
@@ -78,3 +78,289 @@ fn _simply_matches_value_place(left: &PatternValuePlace, right: &PatternValuePla
+
+                // A query might be:
+                // [:find ?x :where (or (and [?x _ 5] [?x :foo/bar 7]))]
+                // which is equivalent to dropping the `or` _and_ the `and`!
+                OrWhereClause::And(clauses) => {
+                    for clause in clauses {
+                        self.apply_clause(schema, clause)?;
+                    }
+                    Ok(())
+                },
+            },
+            _ => self.apply_non_trivial_or_join(schema, or_join),
+        }
+    }
+
+    /// Find out if the `OrJoin` is simple. A simple `or` is one in
+    /// which:
+    /// - Every arm is a pattern, so that we can use a single table alias for all.
+    /// - Each pattern should run against the same table, for the same reason.
+    /// - Each pattern uses the same variables. (That's checked by validation.)
+    /// - Each pattern has the same shape, so we can extract bindings from the same columns
+    ///   regardless of which clause matched.
+    ///
+    /// Like this:
+    ///
+    /// ```
+    /// [:find ?x
+    ///  :where (or [?x :foo/knows "John"]
+    ///             [?x :foo/parent "Ámbar"]
+    ///             [?x :foo/knows "Daphne"])]
+    /// ```
+    ///
+    /// While we're doing this diagnosis, we'll also find out if:
+    /// - No patterns can match: the enclosing CC is known-empty.
+    /// - Some patterns can't match: they are discarded.
+    /// - Only one pattern can match: the `or` can be simplified away.
+    fn deconstruct_or_join(&self, schema: &Schema, mut or_join: OrJoin) -> DeconstructedOrJoin {
+        // If we have explicit non-maximal unify-vars, we *can't* simply run this as a
+        // single pattern --
+        // ```
+        // [:find ?x :where [?x :foo/bar ?y] (or-join [?x] [?x :foo/baz ?y])]
+        // ```
+        // is *not* equivalent to
+        // ```
+        // [:find ?x :where [?x :foo/bar ?y] [?x :foo/baz ?y]]
+        // ```
+        if !or_join.is_fully_unified() {
+            // It's complex because we need to make sure that non-unified vars
+            // mentioned in the body of the `or-join` do not unify with variables
+            // outside the `or-join`. We can't naïvely collect clauses into the
+            // same CC. TODO: pay attention to the unify list when generating
+            // constraints. Temporarily shadow variables within each `or` branch.
+            return DeconstructedOrJoin::Complex(or_join);
+        }
+
+        match or_join.clauses.len() {
+            0 => DeconstructedOrJoin::KnownSuccess,
+
+            // It's safe to simply 'leak' the entire clause, because we know every var in it is
+            // supposed to unify with the enclosing form.
+            1 => DeconstructedOrJoin::Unit(or_join.clauses.pop().unwrap()),
+            _ => self._deconstruct_or_join(schema, or_join),
+        }
+    }
+
+    fn _deconstruct_or_join(&self, schema: &Schema, mut or_join: OrJoin) -> DeconstructedOrJoin {
```

Comment:
A little note explaining the goal of this might help, or just changing `deconstruct` to something that's more like `simplify_...` or `to_simple_...`, since that's what you're really doing: trying to get a simple join at the end.